### PR TITLE
feat: validate commit message for 1 commit PRs

### LINKED ISSUE #86: Validating the commit message for single commit PRs
I understand the scope of this action relates only to PR titles and there are alternatives for those wishing to lint commit messages in general. However, I wonder if you'd consider an optional extra validation step for PRs with only one commit.

When squashing and merging these PRs, GitHub suggests the commit message instead of the PR title, and if you aren't careful the PR will be merged without a conventional commit causing semantic release to ignore it.

I'm happy to put in a PR for this as I'd use it at work.

My suggestion would be an extra boolean configuration option `validateSingleCommit`. If the PR title passes validation and `validateSingleCommit` is true, we get the list of commits and pass if `length > 1`. Otherwise validate the commit message.

### PR DESCRIPTION
**Pull request by simlrh**
Adds an option to validate the commit message when a PR only contains one commit. Closes #86 

When a PR only contains one commit, GitHub suggests that commit message for the "Squash and merge" commit instead of the PR title. This option fails the check if the commit message isn't a conventional commit, preventing an accidental merge of a non-conventional commit into the main branch.

[Here's the PR from this branch to the fork's main branch](https://github.com/simlrh/action-semantic-pull-request/pull/1). It passes since the option is disabled by default.

Demo PRs with the option enabled:

1. [A single commit PR with a non-conventional commit message](https://github.com/simlrh/action-semantic-pull-request/pull/2). Fails.
2. [A single commit PR with a conventional commit message](https://github.com/simlrh/action-semantic-pull-request/pull/3). Passes.
3. [A PR with multiple non-conventional commits](https://github.com/simlrh/action-semantic-pull-request/pull/4). Passes.
4. [For completeness, a PR with a valid single commit but an invalid title](https://github.com/simlrh/action-semantic-pull-request/pull/5). Fails.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -71,6 +71,11 @@ The action works without configuration, however you can provide options for cust
           # validation of the PR title and the pull request checks remain pending.
           # Note that a second check will be reported if this is enabled.
           wip: true
+          # When using "Squash and merge" on a PR with only one commit, GitHub
+          # will suggest using that commit message instead of the PR title for the
+          # merge commit, and it's easy to commit this by mistake. Enable this option
+          # to also validate the commit message for one commit PRs.
+          validateSingleCommit: true
```
Comment (amannn) [inline README.md:78]: Thank you for including good docs!

File: src/index.js [hunk 2]
```diff
@@ -41,13 +42,36 @@ module.exports = async function run() {
...
+            await validateCommitMessage(
+              commits[0].commit.message,
+              {
+                types,
+                scopes,
+                requireScope,
+                subjectPattern,
+                subjectPatternError
+              },
+              'single commit message'
```
Comment (amannn) [inline src/index.js:71]: Nit: I'd move the `messageType` inside the second options argument. I'm using this object as a workaround for named arguments as I think it helps a bit with clarity, making it clear at the call site what the argument represents.

But the comment below is closely related to this as well.

File: src/validateCommitMessage.test.js [hunk 3]
```diff
@@ -174,8 +176,16 @@ describe('description validation', () => {
...
       subjectPattern: '^(?![A-Z]).+$'
     });
   });
 });
+
+it('throws an appropriate error message for single commit message errors', async () => {
+  await expect(
+    validateCommitMessage('Fix bug', undefined, 'single commit message')
+  ).rejects.toThrow(
+    'No release type found in single commit message "Fix bug".'
```
Comment (amannn) [inline src/validateCommitMessage.test.js:189]: Currently the full error message looks like this:

> No release type found in single commit message "Validate commit message for 1 commit PRs (and test option)". Add a prefix to indicate what kind of release this pull request corresponds to (see https://www.conventionalcommits.org/).

I'm not sure this helps the PR author to address the issue since we say "Add a prefix". 

Maybe we should clarify that we're working around an issue of Github here, possibly referencing https://github.community/t/how-to-change-the-default-squash-merge-commit-message/1155 and telling the user to add another commit to avoid the issue? That's also what semantic-pull-requests does.

Since the error message becomes a bit different here, maybe it makes sense to pass in an enum value to `validateCommitMessage` that can be identified and used to alter the message as necessary.

E.g.
```js
validateMessage('foo', {type: 'title'})
validateMessage('foo', {type: 'commit'})
```

I've also renamed to function again here, since it can be either a commit message or a PR title.

What do you think?

**Review by amannn — COMMENTED** [review]
Hi @simlrh,

thank you for starting the discussion around this issue and this corresponding pull request. I noticed this behaviour of Github as well and was a bit put off by it.

I was initially a bit opposed to adding this, similar to how you mentioned, since it increases the scope of the action a bit, but since you made it opt-in, I guess it's just a pragmatic workaround.

I also saw that a similar tool like this, semantic-pull-requests, also discussed this issue in https://github.com/zeke/semantic-pull-requests/issues/17 and fixed it in https://github.com/zeke/semantic-pull-requests/pull/105.

FWIW this seems to be addressed in refined-github: https://github.com/sindresorhus/refined-github/pull/423, but it might be difficult to force all reviewing maintainers of a given repository to use that extension.

So long story short, I think this is a good addition! Thank you so much for taking the time to implement it carefully 👏.

Please consider the two comments I added.

File: src/validateCommitMessage.test.js (hunk 3 above)
Comment (simlrh) [reply src/validateCommitMessage.test.js:189]: I don't want to complicate the logic in `validateMessage` too much - how about I revert the changes I made there, and use a `try/catch` in `index.js` to replace any single commit error with a catch-all message for the problem?

Comment (amannn) [reply src/validateCommitMessage.test.js]: That's a great idea!

**Review by amannn — APPROVED** [review]
